### ISSUE TITLE
Semantically equivalent DAE models give different results

### ISSUE THREAD

**Issue by mtcooling** [opencor/opencor#1321]
This is probably more to do with the integrator than OpenCOR, but

Below are models that seem to be semantically equivalent, but they give different results.
In those models, A and B can either have the values 1 or 2 (C is just there to provide something to integrate over). The piecewises are set up so that when one variable =1, the other = 2. This only happens for one of the models ('problem2') below.

```
def model problem1 as
    def comp all as
        var t: dimensionless;
        var A: dimensionless {init: 1};
        var B: dimensionless {init: 2};
        var C: dimensionless {init: 0};

        ode(C, t) = 1{dimensionless};

        A = sel
            case B == 2{dimensionless}:
                2{dimensionless};
            otherwise:
                1{dimensionless};
        endsel;

        B = sel
            case A == 2{dimensionless}:
                2{dimensionless};
            otherwise:
                1{dimensionless};
        endsel;
    enddef;
enddef;
```

and

```
def model problem2 as
    def comp all as
        var t: dimensionless;
        var A: dimensionless {init: 1};
        var B: dimensionless {init: 2};
        var C: dimensionless {init: 0};

        ode(C, t) = 1{dimensionless};

        A = sel
            case B == 2{dimensionless}:
                2{dimensionless};
            otherwise:
                1{dimensionless};
        endsel;

        B = sel
            case A == 1{dimensionless}:
                1{dimensionless};
            otherwise:
                2{dimensionless};
        endsel;
    enddef;
enddef;
```

**Comment by agarny** [comment]
Nice DAE models! They are definitely useful to test/check OpenCOR. Now, when it comes to `problem2`, it does indeed look like it has more to do with the integrator than OpenCOR itself.

However, please keep in mind that OpenCOR is still very much experimental when it comes to solving DAE models. So, although it looks like OpenCOR can solve some DAE models, there are definitely DAE models that it cannot solve (e.g. see issues #188 and #536).

In other words, we cannot conclude anything at this stage. I first need to get OpenCOR to handle DAE models properly. (Unfortunately, the priority is not currently on being able to solve DAE models...)

**Comment by agarny** [comment]
I have just had another look and although the two problems are indeed semantically equivalent, they are not numerically equivalent. Indeed, the system of two algebraic equations is being solved using floats, not integers. (FWIW, if the system was solved using integers, the solution wouldn't converge since, at a given iteration, `A` and `B` would be equal to either `1` and `2`, or `2` and `1`, respectively.)

In `problem1`, we have to solve:

```
A = (B == 2)?2:1
B = (A == 2)?2:1
```

To do this, we need to compute the difference between the old and new values of `A` and `B`, i.e.

```
diffA = A - (B == 2)?2:1
diffB = B - (A == 2)?2:1
```

Starting with `A = 1` and `B = 2`, we get:

```
diffA = -1
diffB = 1
```

If we consider a solver step of `0.1`, this means that the new values of `A` and `B` are:

```
A = 1.1
B = 1.9
```

If we do this for `100` iterations, we end up with (see [Issue #1321.xlsx](https://github.com/opencor/opencor/files/1596192/Issue.1321.xlsx)):

Iteration # | A | B
-----|---|---
N/A | 1 | 2
1 | 1.1 | 1.9
2 | 1.09 | 1.81
3 | 1.081 | 1.729
... | ... | ...
100 | 1.000002951 | 1.000026561

Similarly, for `problem2`, we end up with:

Iteration # | A | B
-----|---|---
N/A | 1 | 2
1 | 1.1 | 1.9
2 | 1.09 | 1.91
3 | 1.081 | 1.919
... | ... | ...
100 | 1.000002951 | 1.999997049

In other words, everything is actually as expected, hence closing.
